### DISCUSSION TITLE
An input rule for converting [...](url) into links?

### DISCUSSION THREAD

**Discussion by neongreen** [ueberdosis/tiptap#1865] [General Questions]
Does anybody have an example of an input rule that would convert Markdown urls (`[text](url)`) into links?

I am unfamiliar with input rules and I'd appreciate an example like that a lot, even if it does something a bit different.

**Comment by cadars** [comment]
I use this, taken from the [Gitlab editor](https://gitlab.com/gitlab-org/gitlab/-/blob/master/app/assets/javascripts/content_editor/extensions/link.js):

```js
const markdownLinkSyntaxInputRuleRegExp = /(?:^|\s)\[([\w|\s|-]+)\]\((?<href>.+?)\)$/gm;
const urlSyntaxRegExp = /(?:^|\s)(?<href>(?:https?:\/\/|www\.)[\S]+)(?:\s|\n)$/gim;

const extractHrefFromMatch = (match) => {
  return { href: match.groups.href };
};

const extractHrefFromMarkdownLink = (match) => {
  match.pop();
  return extractHrefFromMatch(match);
};

const CustomLink = Link.extend({
  addInputRules() {
    return [
      markInputRule(markdownLinkSyntaxInputRuleRegExp, this.type, extractHrefFromMarkdownLink),
      markInputRule(urlSyntaxRegExp, this.type, extractHrefFromMatch),
    ];
  },
});
```

**Reply by hanspagel** [reply]
@philippkuehn Do you think we should add this to the link Extension?

**Reply by neongreen** [reply]
One more thing — this doesn't play well with paste rules. When writing `[foo](` and then pasting a link, it becomes autolinked and then the `[]()` input rule doesn't fire. I've spent about an hour trying to fix this, but `markPasteRule` doesn't seem to be powerful enough to allow detecting `](`.

**Reply by cadars** [reply]
> @philippkuehn Do you think we should add this to the link Extension?

I really like this suggestion :) I understand that Markdown input is added as an extra convenience in tiptap, but it would be awesome to be able to simply input links without a menu, prompt, or more complicated components.

**Comment by cadars** [comment]
I rewrote the input rules following [this change](https://github.com/ueberdosis/tiptap/pull/1997), and it now seems to work when pasting links within `()`

```js
const CustomLink = Link.extend({
  addInputRules() {
    return [
      markInputRule({
        find: markdownLinkSyntaxInputRuleRegExp,
        type: this.type,
        getAttributes: extractHrefFromMarkdownLink,
      }),
      markInputRule({
        find: urlSyntaxRegExp,
        type: this.type,
        getAttributes: extractHrefFromMatch,
      }),
    ];
  },
});
```
I guess it would be better to use linkify in some way to get the `href`

**Reply by cadars** [reply]
update: the new autolink feature (https://github.com/ueberdosis/tiptap/pull/2226) means that the code above won’t work anymore, I might try to rewrite it, but if anyone has any pointers…

**Reply by hanspagel** [reply]
Just so you know, you can disable the autolink feature in the link extension. ✌️

**Reply by cadars** [reply]
That works, thanks! I'll try to fix it anyway, autolink is really nice to have. My goal is to have Zettlr-style links, where clicking an existing link expands the markdown syntax for it, and re-renders the link when focusing somewhere else… exactly how prosemirror-math does for inline math.

**Comment by JonathanKolnik** [comment]
@cadars really appreciate the help here. If it's not too much trouble could you explain how this works a little further (just for my own learnings). I understand how the regex matches for this syntax but how is this actually updating the contents of the mark?

**Reply by cadars** [reply]
Being honest, I couldn’t say, I just copied the Gitlab link extension: https://gitlab.com/gitlab-org/gitlab/-/blob/master/app/assets/javascripts/content_editor/extensions/link.js

**Comment by justinmoon** [comment]
The method mentioned by @cadars is working nicely for me. I do have one questions.

If I put my cursor after the last character of a link and hit the "delete" key, I'd like for the `Link` mark to disappear and be replaced by plaintext `[description](link.com)` so that I could manually edit it again. Instead, currently, it just deletes the last character of the link description if that makes sense.

How might I do this?

**Comment by LoiKos** [comment]
Hi guys, 

I have dig a little bit into tiptap code to try to figure out a solution for this. 

After debugging, i found out that it was `autolink` that was removing the newly created link because the text was not evaluted as a valid link. see : 
https://github.com/ueberdosis/tiptap/blob/ab4a0e2507b4b92c46d293a0bb06bb00a04af6e0/packages/extension-link/src/helpers/autolink.ts#L46-L56

It seems there is an easy solution to this problem as we can prevent using `autolink` but this solution required to modify  :

https://github.com/ueberdosis/tiptap/blob/ab4a0e2507b4b92c46d293a0bb06bb00a04af6e0/packages/core/src/inputRules/markInputRule.ts#L62-L65

by adding :

```ts
        tr.setMeta('preventAutolink', true)
```       
      
Could it be a solution acceptable for you ? 

I'll be happy to submit a PR if this is fine for you ! 

**This is potentially breaking for some users because autolink won't work with** `markInputRule`
I wish i could propose a solution with a per-markInputRule flag and only in case of Link. 
I don't think it's possible or i don't know how to do it.

**Comment by LoiKos** [comment]
Finally, i found a way to handle the markdown link keeping autolink activated. It's really close to above solution :

```typescript
import Link from '@tiptap/extension-link'
import {
  ExtendedRegExpMatchArray,
  InputRule,
  markInputRule,
} from '@tiptap/react'

const markdownLinkSyntaxInputRuleRegExp =
  /(?:^|\s)\[([\w|\s|-]+)\]\((?<href>.+?)\)$/gm

const extractHrefFromMatch = (match: ExtendedRegExpMatchArray) => {
  return { href: match.groups?.href }
}

const extractHrefFromMarkdownLink = (match: ExtendedRegExpMatchArray) => {
  match.pop()
  return extractHrefFromMatch(match)
}

let linkInputRule = (config: Parameters<typeof markInputRule>[0]) => {
  let defaultMarkInputRule = markInputRule(config)

  return new InputRule({
    find: config.find,
    handler: (props) => {
      defaultMarkInputRule.handler(props)
      const { tr } = props.state
      tr.setMeta('preventAutolink', true)
    },
  })
}

export default Link.extend({
  addInputRules() {
    return [
      linkInputRule({
        find: markdownLinkSyntaxInputRuleRegExp,
        type: this.type,
        getAttributes: extractHrefFromMarkdownLink,
      }),
    ]
  },
})
```

**Comment by rfgamaral** [comment]
Based on everyone's solution above, and especially the auto-link fix by @LoiKos (thank you so much 🙏), I've ended up with the following solution for us:

```ts
import { InputRule, markInputRule, markPasteRule, PasteRule } from '@tiptap/core'
import { Link as TiptapLink } from '@tiptap/extension-link'

import type { LinkOptions } from '@tiptap/extension-link'

/**
 * The input regex for Markdown links with title support, and multiple quotation marks (required
 * in case the `Typography` extension is being included).
 *
 * @see https://stephenweiss.dev/regex-markdown-link
 */
const inputRegex = /(?:^|\s)\[([^\]]*)?\]\(([A-Za-z0-9:/. ]+)(?:["“](.+)["”])?\)$/

/**
 * The paste regex for Markdown links with title support, and multiple quotation marks (required
 * in case the `Typography` extension is being included).
 *
 * @see https://stephenweiss.dev/regex-markdown-link
 */
const pasteRegex = /(?:^|\s)\[([^\]]*)?\]\(([A-Za-z0-9:/. ]+)(?:["“](.+)["”])?\)/g

/**
 * Input rule built specifically for the `Link` extension, which ignores the auto-linked URL in
 * parentheses (e.g., `(https://doist.dev)`).
 *
 * @see https://github.com/ueberdosis/tiptap/discussions/1865
 */
function linkInputRule(config: Parameters<typeof markInputRule>[0]) {
    const defaultMarkInputRule = markInputRule(config)

    return new InputRule({
        find: config.find,
        handler: (props) => {
            const { tr } = props.state

            defaultMarkInputRule.handler(props)
            tr.setMeta('preventAutolink', true)
        },
    })
}

/**
 * Paste rule built specifically for the `Link` extension, which ignores the auto-linked URL in
 * parentheses (e.g., `(https://doist.dev)`).
 *
 * @see https://github.com/ueberdosis/tiptap/discussions/1865
 */
function linkPasteRule(config: Parameters<typeof markPasteRule>[0]) {
    const defaultMarkInputRule = markPasteRule(config)

    return new PasteRule({
        find: config.find,
        handler: (props) => {
            const { tr } = props.state

            defaultMarkInputRule.handler(props)
            tr.setMeta('preventAutolink', true)
        },
    })
}

/**
 * Custom extension that extends the built-in `Link` extension to add additional input and paste
 * rules for converting the Markdown link syntax [Doist](https://doist.com) into links. This
 * extension also adds support for the `title` attribute.
 */
const Link = TiptapLink.extend({
    addAttributes() {
        return {
            ...this.parent?.(),
            title: {
                default: null,
            },
        }
    },
    addInputRules() {
        return [
            linkInputRule({
                find: inputRegex,
                type: this.type,

                // We need to use `pop()` to remove the last capture groups from the match to
                // satisfy Tiptap's `markPasteRule` expectation of having the content as the last
                // capture group in the match (this makes the attribute order important)
                getAttributes(match) {
                    return {
                        title: match.pop()?.trim(),
                        href: match.pop()?.trim(),
                    }
                },
            }),
        ]
    },
    addPasteRules() {
        return [
            linkPasteRule({
                find: pasteRegex,
                type: this.type,

                // We need to use `pop()` to remove the last capture groups from the match to
                // satisfy Tiptap's `markInputRule` expectation of having the content as the last
                // capture group in the match (this makes the attribute order important)
                getAttributes(match) {
                    return {
                        title: match.pop()?.trim(),
                        href: match.pop()?.trim(),
                    }
                },
            }),
        ]
    },
})

export { Link }

export type { LinkOptions }
```

The main differences are:

* The regex was modified/adapted with support for link titles (and it supports double quotes converted from the `Typography` extension), and I also removed the named capturing groups (although this is nice and all, some "not so old" Firefox version don't support his, and we had users complaining that the app completely broke for them)
* As alluded from the previous point, the extended extension also supports the `title` attribute for `<a>` elements (which is missing from the official extension)
